Support `-Cforce-frame-pointers=non-leaf`

workingjubilee opened the pull request:
Why don't we already support this...?

Suggested impl for https://github.com/rust-lang/compiler-team/issues/744

workingjubilee committed:
compiler: Add FramePointer::ratchet

workingjubilee committed:
compiler(nfc): -Cforce-frame-pointers is a FramePointer

workingjubilee committed:
cg_clif: Define build opts from FramePointer

workingjubilee committed:
compiler: Support nightly -Cforce-frame-pointers=non-leaf

Requires -Zunstable-options as this is a -C flag already.

workingjubilee committed:
test: for frame-pointer=non-leaf codegen opts

Zalathar commented on compiler/rustc_session/src/config.rs:
```diff
@@ -2468,6 +2468,15 @@ pub fn build_session_options(early_dcx: &mut EarlyDiagCtxt, matches: &getopts::M
         }
     }
 
+    if !nightly_options::is_unstable_enabled(matches)
```
Annoyingly, there seem to be multiple different ways of checking for `-Zunstable-options` when implementing these ad-hoc value checks:

- Some code uses `nightly_options::is_unstable_enabled(matches)`, which parses the value independently
- Some code uses `unstable_opts.unstable_options`, which uses whatever has already been parsed into the `UnstableOptions` struct via normal means

In this context, we should be able to just check `unstable_opts.unstable_options`, which seems like the less hacky variant.

(I assume you probably just copied this from the code immediately below. 🤷)

workingjubilee replied:
I did check the unstable_options.unstable_options thing, but it says you shouldn't touch it directly, so I avoided poking it directly. idk lol?

I think there should be a different way of specifying "these values are unstable" that uses a pattern or pattern-guard of some kind.

workingjubilee replied:
I realize there's only a few cases of this so it seems like a waste, but I think a profusion of flags is partly the fault of it not being easy to add such patterns that determine stability.

Zalathar replied:
I don’t want to derail this PR too much, but if I get a chance I’ll open a separate issue or thread to discuss improvements to unstable values of stable flags, because the status quo has been annoying me too.

jsgf commented on compiler/rustc_session/src/options.rs:
```diff
@@ -670,6 +671,18 @@ mod parse {
     }
 
+    pub(crate) fn parse_frame_pointer(slot: &mut FramePointer, v: Option<&str>) -> bool {
+        let mut boolish = false;
+        let mut is_parsed = parse_bool(&mut boolish, v);
+        if boolish & is_parsed {
+            *slot = FramePointer::Always;
+        } else if v == Some("non-leaf") {
+            is_parsed = true;
+            *slot = FramePointer::NonLeaf;
+        };
```
I think this would be a bit cleaner as a `match`.

workingjubilee replied:
Sure!

jsgf commented on compiler/rustc_session/src/config.rs:
```diff
@@ -2468,6 +2468,15 @@ pub fn build_session_options(early_dcx: &mut EarlyDiagCtxt, matches: &getopts::M
+    if !nightly_options::is_unstable_enabled(matches)
+        && cg.force_frame_pointers == FramePointer::NonLeaf
+    {
+        early_dcx.early_fatal(
+            "`-Cforce-frame-pointers=non-leaf` also requires `-Zunstable-options` \
+                and a nightly compiler",
+        )
```
Do we really need this? What's the concern? It's already de-facto enabled via `--target=spec.json`.

workingjubilee replied:
Target specs are also unstable.

workingjubilee replied:
Mind, we _could_ land it as instantly stable, I agree it has strong precedent, but then it needs to wait on a full T-compiler FCP, so "this can happen **right now** without requiring everyone to agree" was why I proposed doing things "correctly".

jsgf commented on compiler/rustc_session/src/options.rs:
```diff
@@ -670,6 +671,18 @@ mod parse {
         }
     }
 
+    pub(crate) fn parse_frame_pointer(slot: &mut FramePointer, v: Option<&str>) -> bool {
```
Maybe `parse_force_frame_pointers` so that it matches the actual option?

workingjubilee replied:
The other parse options reflect the value they parse.

jsgf commented on compiler/rustc_target/src/spec/mod.rs:
```diff
@@ -1385,6 +1385,20 @@ pub enum FramePointer {
 }
 
+impl FramePointer {
+    /// It is intended that the "force frame pointer" transition is "one way"
+    /// so this convenience assures such if used
+    #[inline]
+    pub fn ratchet(&mut self, rhs: FramePointer) -> FramePointer {
+        *self = match (*self, rhs) {
+            (FramePointer::Always, _) | (_, FramePointer::Always) => FramePointer::Always,
+            (FramePointer::NonLeaf, _) | (_, FramePointer::NonLeaf) => FramePointer::NonLeaf,
+            _ => FramePointer::MayOmit,
+        };
+        *self
+    }
```
This seems a bit over-complex for the one use it has.

workingjubilee replied:
Its real purpose is to document that frame pointer forcing should go up and not down. I also originally used it in the parse, I suppose that got lost at some point.

jsgf replied:
Are "up" and "down" defined for this? I assume "up" means "more frame pointers". But I don't think that's really what we want - we want it to default to whatever the spec sets, and allow it to be overridden by the command line. If the spec says "always" but the command line is "may omit" (or always or non-leaf) then that should always take preference.

workingjubilee replied:
https://github.com/rust-lang/rust/blob/9e7aff794539aa040362f4424eb29207449ffce0/src/doc/rustc/src/codegen-options/index.md?plain=1#L157-L165

workingjubilee replied:
Removing frame pointers can break the standard library's runtime behaviors. A given target can require frame pointers and simply stop working correctly (e.g. backtraces or unwinding may stop working) if software is compiled otherwise. While not everyone values backtraces much (else this compilation option would never have come about), breaking the unwinder seems... bad.

jsgf replied:
This is an advanced option for people who know what they're doing. I think it's important that it do exactly what it claims to do. Obviously it shouldn't generate code that's so broken that it's non-viable, but to the greatest extent possible it should allow overriding the default setting for that target. There are lots of cases where backtraces/unwinding might be considered secondary to whatever goal disabling frame pointers is trying to achieve and are therefore expendible.

>  breaking the unwinder seems... bad.

That's moot if you're using panic=abort. And in many (most? all?) cases unwinding does not rely on frame pointers.

bjorn3 replied:
On macOS the arm64 ABI mandates the existence of frame pointers. On x86_64 frame pointers are also mandated with panic=unwind: https://github.com/rust-lang/rust/issues/74640

On Windows breaking the unwinder will cause problems even with panic=abort as the OS throws a SEH exception whenever the stack needs to grow. This indeed doesn't depend on frame pointers though.

workingjubilee replied:
@jsgf It currently does do what it says it does, per the docs. Changing the behavior the way you describe is a much more significant change and one I do not think we should try to smuggle into a much smaller proposal to add a flag.

jsgf replied:
Right now "no" means "MayOmit", which leaves it up to llvm to decide how to behave - this is fine. It still allows the target spec to be overridden.

The way `ratchet` is implemented means that the spec always takes precedence over the command line option if its "stronger". I want to be able to override this. It's also a change of behaviour compared to the current code. Right now, the only thing that always forces frame pointers is using mcount.

workingjubilee replied:
@jsgf It does not amount to a change in behavior! *Please*, reread the function, the version before I touched it:

https://github.com/rust-lang/rust/blob/6e5f1dacf3a670cf03111039a032c765355839aa/compiler/rustc_codegen_llvm/src/attributes.rs#L106-L120

On line 107, we get this code which mutably declares `fp`:
https://github.com/rust-lang/rust/blob/6e5f1dacf3a670cf03111039a032c765355839aa/compiler/rustc_codegen_llvm/src/attributes.rs#L107

That's the Session's target, which is derived from the spec. We get the field `frame_pointer` from the spec, right there. That is the spec's value. That means it is what we default to if no other code changes it, yes?

On line 111, we get this code:
https://github.com/rust-lang/rust/blob/6e5f1dacf3a670cf03111039a032c765355839aa/compiler/rustc_codegen_llvm/src/attributes.rs#L111

There are two conditions there. One is `mcount`. The other is *also* from the session. It's the session's parsed configuration value for the `force_frame_pointers` option. It is `Some(true)` if you set `-Cforce-frame-pointers=yes` (or any of the other positive values).

These conditions are joined by `||`, a short-circuiting `or`, so only one must be `true` in order for us to reach the next line:
https://github.com/rust-lang/rust/blob/6e5f1dacf3a670cf03111039a032c765355839aa/compiler/rustc_codegen_llvm/src/attributes.rs#L111
...where we set `FramePointer::Always`.

That's the only possible change. The fact that we accept `-Cforce-frame-pointers=no` on the command line is a cruel trick, in effect. It won't actually do anything.

The "target" does in fact take precedence.

workingjubilee replied:
_Perhaps_ some subtlety has amounted to a change *elsewhere*... we're modifying code in multiple places, after all... but I believe the `frame_pointer_type_attr`, the one that combines the various configuration options into a single "set frame pointers?" value, continues to behave exactly as it did before. For all the currently-settable values that are not being introduced in this PR, anyways. I believe the behavior you dislike is the current behavior.

I believe the code, unfortunately, is now clearer as to how it functions, and no longer misleads you as to what happens. Perhaps that has led you to dislike it? That seems a reasonable response, and I am not trying to persuade you that your requested behavior is unreasonable. I _am_ trying to persuade you that your request is _not what is implemented currently_, and that there are at least _reasons to make it a separate process_.

I think it amounts to promising a new behavior, basically, which as I understand it requires a T-compiler FCP? Again, "that would require more process" is a reason I went for drafting the PR I currently did.

The _new_ thing is the `-Cforce-frame-pointers=non-leaf` setting. I implemented it the way I have because this causes it to align _notionally_ with the way the code currently works: it takes the highest value between the one in the spec, the one given on the command line, and the one required by `instrument_mcount`.

jsgf replied:
Yeah sorry I re-read the original and I agree that spec overrides. Well, I guess we can still use json specs if we really need to.

workingjubilee replied:
@jsgf I notice your description of the "current behavior" accurately describes the implemented behavior in https://github.com/rust-lang/rust/pull/125127 which, as I mention, unless I am horribly wrong, is different from the current behavior of `frame_pointer_type_attr` in the compiler. I chose not to implement things that way for all the reasons I've already described. Namely: it introduces new behavior that is not how the code currently works.

Please take this line of inquiry to a new *and unrelated* issue, PR, etc.

workingjubilee replied:
Thank you.

workingjubilee commented:
https://github.com/rust-lang/rust/pull/125127 was opened which suggests "always" and "never" as other options. I think adding "always" to this is fine if there's no objection from the reviewer (who also seconded the MCP) but I don't believe we need more options for saying "no" because "never" is even more confusing than "no" or "off" (because it is strong enough it might come off as "force frame pointers to NEVER HAPPEN!!!" which is not at all what will occur).

workingjubilee committed:
compiler: Accept -Cforce-frame-pointers=always

Also lands behind -Zunstable-options, for now.
Take the opportunity to do some mild cleanup.

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9085256881/job/24968155650) [(plain)](https://github.com/rust-lang/rust/commit/f5ad1164ef68671277aae03b74565507e5fcb0e0/checks/24968155650/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Getting action download info
Download action repository 'msys2/setup-msys2@v2.22.0' (SHA:cc11e9188b693c2b100158c3322424c4cc1dadea)
Download action repository 'actions/checkout@v4' (SHA:0ad4b8fadaa221de15dcec353f45205ec38ea70b)
Download action repository 'actions/upload-artifact@v4' (SHA:65462800fd760344b1a7b4382951275a0abb4808)
Complete job name: PR - mingw-check-tidy
git config --global core.autocrlf false
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
---
COPY scripts/sccache.sh /scripts/
RUN sh /scripts/sccache.sh

COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt \
    && pip3 install virtualenv
COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---

#12 [5/8] COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
#12 DONE 0.0s

#13 [6/8] RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt     && pip3 install virtualenv
#13 0.427   Downloading binaryornot-0.4.4-py2.py3-none-any.whl (9.0 kB)
#13 0.444 Collecting boolean-py==4.0
#13 0.451   Downloading boolean.py-4.0-py3-none-any.whl (25 kB)
#13 0.471 Collecting chardet==5.1.0
---
#13 3.624 Building wheels for collected packages: reuse
#13 3.624   Building wheel for reuse (pyproject.toml): started
#13 3.948   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 3.949   Created wheel for reuse: filename=reuse-1.1.0-cp310-cp310-manylinux_2_35_x86_64.whl size=181117 sha256=f5f58750481f69515c2c0d1d503daf565e2565c370d07fc6aeb95fe3498b4269
#13 3.949   Stored in directory: /tmp/pip-ephem-wheel-cache-ivpzbrix/wheels/c2/3c/b9/1120c2ab4bd82694f7e6f0537dc5b9a085c13e2c69a8d0c76d
#13 3.952 Installing collected packages: boolean-py, binaryornot, setuptools, reuse, python-debian, markupsafe, license-expression, jinja2, chardet
#13 3.973   Attempting uninstall: setuptools
#13 3.974     Found existing installation: setuptools 59.6.0
#13 3.975     Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
---
#13 5.252   Downloading virtualenv-20.26.2-py3-none-any.whl (3.9 MB)
#13 5.310      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.9/3.9 MB 71.3 MB/s eta 0:00:00
#13 5.364 Collecting filelock<4,>=3.12.2
#13 5.372   Downloading filelock-3.14.0-py3-none-any.whl (12 kB)
#13 5.404 Collecting platformdirs<5,>=3.9.1
#13 5.412   Downloading platformdirs-4.2.1-py3-none-any.whl (17 kB)
#13 5.433 Collecting distlib<1,>=0.3.7
#13 5.441   Downloading distlib-0.3.8-py2.py3-none-any.whl (468 kB)
#13 5.447      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 468.9/468.9 KB 119.0 MB/s eta 0:00:00
#13 5.532 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 5.694 Successfully installed distlib-0.3.8 filelock-3.14.0 platformdirs-4.2.1 virtualenv-20.26.2
#13 DONE 5.8s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      210880 kB
DirectMap2M:     7129088 kB
DirectMap1G:    11534336 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
    Finished `dev` profile [unoptimized] target(s) in 0.03s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/ac385a5af6d9fa8399a0cec799833cd28324abf8/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-ac385a5af6d9fa8399a0cec799833cd28324abf8-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
---
    Finished `release` profile [optimized] target(s) in 25.17s
##[endgroup]
fmt check
tidy check
tidy error: unknown revision [Always] at /checkout/tests/codegen/force-frame-pointers.rs:2
tidy error: unknown revision [NonLeaf] at /checkout/tests/codegen/force-frame-pointers.rs:3
removing old virtual environment
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (24.0)
Collecting black==23.3.0 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 7))
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.7/1.7 MB 19.9 MB/s eta 0:00:00
Collecting click==8.1.3 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 34))
  Downloading click-8.1.3-py3-none-any.whl (96 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 96.6/96.6 kB 39.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 96.6/96.6 kB 39.6 MB/s eta 0:00:00
Collecting importlib-metadata==6.7.0 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 38))
  Downloading importlib_metadata-6.7.0-py3-none-any.whl (22 kB)
Collecting mypy-extensions==1.0.0 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 42))
  Downloading mypy_extensions-1.0.0-py3-none-any.whl (4.7 kB)
Collecting packaging==23.1 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 46))
  Downloading packaging-23.1-py3-none-any.whl (48 kB)
Collecting pathspec==0.11.1 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 50))
  Downloading pathspec-0.11.1-py3-none-any.whl (29 kB)
Collecting platformdirs==3.6.0 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 54))
  Downloading platformdirs-3.6.0-py3-none-any.whl (16 kB)
  Downloading platformdirs-3.6.0-py3-none-any.whl (16 kB)
Collecting ruff==0.0.272 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 58))
  Downloading ruff-0.0.272-py3-none-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (5.9 MB)
Collecting tomli==2.0.1 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 77))
  Downloading tomli-2.0.1-py3-none-any.whl (12 kB)
Collecting typed-ast==1.5.4 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 81))
  Downloading typed_ast-1.5.4-cp310-cp310-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_12_x86_64.manylinux2010_x86_64.whl (877 kB)
  Downloading typed_ast-1.5.4-cp310-cp310-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_12_x86_64.manylinux2010_x86_64.whl (877 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 877.7/877.7 kB 66.8 MB/s eta 0:00:00
Collecting typing-extensions==4.6.3 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 107))
  Downloading typing_extensions-4.6.3-py3-none-any.whl (31 kB)
Collecting zipp==3.15.0 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 114))
  Downloading zipp-3.15.0-py3-none-any.whl (6.8 kB)
Installing collected packages: zipp, typing-extensions, typed-ast, tomli, ruff, platformdirs, pathspec, packaging, mypy-extensions, click, importlib-metadata, black
Successfully installed black-23.3.0 click-8.1.3 importlib-metadata-6.7.0 mypy-extensions-1.0.0 packaging-23.1 pathspec-0.11.1 platformdirs-3.6.0 ruff-0.0.272 tomli-2.0.1 typed-ast-1.5.4 typing-extensions-4.6.3 zipp-3.15.0
some tidy checks failed
Build completed unsuccessfully in 0:00:58
  local time: Tue May 14 19:50:20 UTC 2024
  network time: Tue, 14 May 2024 19:50:20 GMT
```

</details>

jsgf commented on compiler/rustc_interface/src/tests.rs:
```diff
@@ -605,7 +605,7 @@ fn test_codegen_options_tracking_hash() {
     tracked!(debuginfo, DebugInfo::Limited);
     tracked!(embed_bitcode, false);
-    tracked!(force_frame_pointers, Some(false));
+    tracked!(force_frame_pointers, FramePointer::Always);
```
Should this be MayOmit?

bjorn3 replied:
Any value other than the default is correct here. This tests that changing the value of the option changes the dep tracking hash.

workingjubilee replied:
and it seemed like it would be a cold day in hell before `FramePointer::Always` was the default.

jsgf replied:
Oh Some(false) vs None

workingjubilee commented:
https://github.com/rust-lang/compiler-team/issues/744 was accepted.

workingjubilee commented:
r? T-compiler

compiler-errors commented:
I have no context about this, sorry

r? compiler

workingjubilee commented:
r? compiler

saethlin commented:
r? saethlin
@bors r+

bors commented:
:pushpin: Commit f705368dead65c33d446b9cdee2897dfb3211826 has been approved by `saethlin`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=saethlin f705368dead65c33d446b9cdee2897dfb3211826 -->
<!-- homu: {"type":"Approved","sha":"f705368dead65c33d446b9cdee2897dfb3211826","approver":"saethlin","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

saethlin commented:
Looks like it works and it's an unstable flag addition. If we are going to stabilize this, people need to try it out. So we send.

bors commented:
:hourglass: Testing commit f705368dead65c33d446b9cdee2897dfb3211826 with merge 61f0811f46883ae57c766608b7f17b312710351b...
<!-- homu: {"type":"BuildStarted","head_sha":"f705368dead65c33d446b9cdee2897dfb3211826","merge_sha":"61f0811f46883ae57c766608b7f17b312710351b"} -->

rust-log-analyzer commented:
The job **`dist-i586-gnu-i586-i686-musl`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/9630264262/job/26560643846) [(plain)](https://github.com/rust-lang-ci/rust/commit/61f0811f46883ae57c766608b7f17b312710351b/checks/26560643846/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
failures:

---- [codegen] tests/codegen/force-frame-pointers.rs#NonLeaf stdout ----

error in revision `NonLeaf`: verification with 'FileCheck' failed
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/llvm/build/bin/FileCheck" "--input-file" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/force-frame-pointers.NonLeaf/force-frame-pointers.ll" "/checkout/tests/codegen/force-frame-pointers.rs" "--check-prefix=CHECK" "--check-prefix" "NonLeaf" "--check-prefix" "NONMSVC" "--allow-unused-prefixes" "--dump-input-context" "100"
--- stderr -------------------------------
--- stderr -------------------------------
/checkout/tests/codegen/force-frame-pointers.rs:10:13: error: NonLeaf: expected string not found in input
// NonLeaf: attributes #{{.*}} "frame-pointer"="non-leaf"
            ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/force-frame-pointers.NonLeaf/force-frame-pointers.ll:1:1: note: scanning from here
; ModuleID = 'force_frame_pointers.51a664ea9b58e609-cgu.0'
^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/force-frame-pointers.NonLeaf/force-frame-pointers.ll:13:12: note: possible intended match here
attributes #0 = { uwtable "frame-pointer"="all" "probe-stack"="inline-asm" "target-cpu"="pentium4" }


Input file: /checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/force-frame-pointers.NonLeaf/force-frame-pointers.ll


-dump-input=help explains the following input dump.
Input was:
<<<<<<
<<<<<<
            1: ; ModuleID = 'force_frame_pointers.51a664ea9b58e609-cgu.0' 
check:10'0     X~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ error: no match found
            2: source_filename = "force_frame_pointers.51a664ea9b58e609-cgu.0" 
check:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
            3: target datalayout = "e-m:e-p:32:32-p270:32:32-p271:32:32-p272:64:64-i128:128-f64:32:64-f80:32-n8:16:32-S128" 
check:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
            4: target triple = "i686-unknown-linux-musl" 
check:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:10'0     ~
            6: ; force_frame_pointers::foo 
check:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
            7: ; Function Attrs: uwtable 
            7: ; Function Attrs: uwtable 
check:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~
            8: define void @_ZN20force_frame_pointers3foo17hbb4393e90892147cE() unnamed_addr #0 { 
check:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:10'0     ~~~~~~~
           10:  ret void 
check:10'0     ~~~~~~~~~~
           11: } 
           11: } 
check:10'0     ~~
           12:  
check:10'0     ~
           13: attributes #0 = { uwtable "frame-pointer"="all" "probe-stack"="inline-asm" "target-cpu"="pentium4" } 
check:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:10'1                ?                                                                                          possible intended match
check:10'0     ~
check:10'0     ~
           15: !llvm.module.flags = !{!0} 
check:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~
           16: !llvm.ident = !{!1} 
check:10'0     ~~~~~~~~~~~~~~~~~~~~
check:10'0     ~
check:10'0     ~
           18: !0 = !{i32 8, !"PIC Level", i32 2} 
check:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           19: !1 = !{!"rustc version 1.81.0-nightly (61f0811f4 2024-06-23)"} 
check:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
------------------------------------------



```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/9630264262/job/26561219117)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/9630264262/job/26561219117","builder_name":"checks-actions"} -->

workingjubilee commented:
Oh, some targets set non-leaf frame pointers, right...

workingjubilee committed:
test: ignore force-frame-pointers test on some targets

all of these currently force stronger frame pointers, and
currently the CLI does not override the target

workingjubilee commented:
@bors r=saethlin

bors commented:
:pushpin: Commit 0ba2c0379c1cf1626415403a8aa1411803972d6a has been approved by `saethlin`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=saethlin 0ba2c0379c1cf1626415403a8aa1411803972d6a -->
<!-- homu: {"type":"Approved","sha":"0ba2c0379c1cf1626415403a8aa1411803972d6a","approver":"saethlin","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9630828774/job/26561856586) [(plain)](https://github.com/rust-lang/rust/commit/0ba2c0379c1cf1626415403a8aa1411803972d6a/checks/26561856586/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
 - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 46)
 - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 49)
 - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 61)
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
test result: ok. 292 passed; 0 failed; 5 ignored; 0 measured; 0 filtered out; finished in 4.59s

##[endgroup]
##[group]Testing stage2 compiletest suite=codegen mode=codegen (x86_64-unknown-linux-gnu)
error: detected unknown compiletest test directive `ignore-illumos forces stronger frame pointers` in /checkout/tests/codegen/force-frame-pointers.rs:6
errors encountered during EarlyProps parsing: /checkout/tests/codegen/force-frame-pointers.rs
errors encountered during EarlyProps parsing
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
Build completed unsuccessfully in 0:15:02
  local time: Sun Jun 23 04:56:51 UTC 2024
```

</details>

workingjubilee commented:
@bors r-

workingjubilee commented:
@bors r=saethlin

bors commented:
:pushpin: Commit a40d8bbf84b2918962975bb7a01a36c1184d660b has been approved by `saethlin`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=saethlin a40d8bbf84b2918962975bb7a01a36c1184d660b -->
<!-- homu: {"type":"Approved","sha":"a40d8bbf84b2918962975bb7a01a36c1184d660b","approver":"saethlin","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit a40d8bbf84b2918962975bb7a01a36c1184d660b with merge fd8d5b90d85953bf23811295c8a3fc631b4e6b26...
<!-- homu: {"type":"BuildStarted","head_sha":"a40d8bbf84b2918962975bb7a01a36c1184d660b","merge_sha":"fd8d5b90d85953bf23811295c8a3fc631b4e6b26"} -->

rust-log-analyzer commented:
The job **`x86_64-apple-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/9631219484/job/26562733354) [(plain)](https://github.com/rust-lang-ci/rust/commit/fd8d5b90d85953bf23811295c8a3fc631b4e6b26/checks/26562733354/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
failures:

---- [codegen] tests/codegen/force-frame-pointers.rs#NonLeaf stdout ----

error in revision `NonLeaf`: verification with 'FileCheck' failed
status: exit status: 1
command: "/Users/runner/work/rust/rust/build/x86_64-apple-darwin/ci-llvm/bin/FileCheck" "--input-file" "/Users/runner/work/rust/rust/build/x86_64-apple-darwin/test/codegen/force-frame-pointers.NonLeaf/force-frame-pointers.ll" "/Users/runner/work/rust/rust/tests/codegen/force-frame-pointers.rs" "--check-prefix=CHECK" "--check-prefix" "NonLeaf" "--check-prefix" "NONMSVC" "--allow-unused-prefixes" "--dump-input-context" "100"
--- stderr -------------------------------
--- stderr -------------------------------
/Users/runner/work/rust/rust/tests/codegen/force-frame-pointers.rs:16:13: error: NonLeaf: expected string not found in input
// NonLeaf: attributes #{{.*}} "frame-pointer"="non-leaf"
            ^
/Users/runner/work/rust/rust/build/x86_64-apple-darwin/test/codegen/force-frame-pointers.NonLeaf/force-frame-pointers.ll:1:1: note: scanning from here
; ModuleID = 'force_frame_pointers.51a664ea9b58e609-cgu.0'
^
/Users/runner/work/rust/rust/build/x86_64-apple-darwin/test/codegen/force-frame-pointers.NonLeaf/force-frame-pointers.ll:13:12: note: possible intended match here
attributes #0 = { uwtable "frame-pointer"="all" "probe-stack"="inline-asm" "target-cpu"="penryn" }


Input file: /Users/runner/work/rust/rust/build/x86_64-apple-darwin/test/codegen/force-frame-pointers.NonLeaf/force-frame-pointers.ll


-dump-input=help explains the following input dump.
Input was:
<<<<<<
<<<<<<
            1: ; ModuleID = 'force_frame_pointers.51a664ea9b58e609-cgu.0' 
check:16'0     X~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ error: no match found
            2: source_filename = "force_frame_pointers.51a664ea9b58e609-cgu.0" 
check:16'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
            3: target datalayout = "e-m:o-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32:64-S128" 
check:16'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
            4: target triple = "x86_64-apple-macosx10.12.0" 
check:16'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:16'0     ~
            6: ; force_frame_pointers::foo 
check:16'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
            7: ; Function Attrs: uwtable 
            7: ; Function Attrs: uwtable 
check:16'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~
            8: define void @_ZN20force_frame_pointers3foo17hbb4393e90892147cE() unnamed_addr #0 { 
check:16'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:16'0     ~~~~~~~
           10:  ret void 
check:16'0     ~~~~~~~~~~
           11: } 
           11: } 
check:16'0     ~~
           12:  
check:16'0     ~
           13: attributes #0 = { uwtable "frame-pointer"="all" "probe-stack"="inline-asm" "target-cpu"="penryn" } 
check:16'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:16'1                ?                                                                                        possible intended match
check:16'0     ~
check:16'0     ~
           15: !llvm.module.flags = !{!0} 
check:16'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~
           16: !llvm.ident = !{!1} 
check:16'0     ~~~~~~~~~~~~~~~~~~~~
check:16'0     ~
check:16'0     ~
           18: !0 = !{i32 8, !"PIC Level", i32 2} 
check:16'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           19: !1 = !{!"rustc version 1.81.0-nightly (fd8d5b90d 2024-06-23)"} 
check:16'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
------------------------------------------



```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/9631219484/job/26563480565)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/9631219484/job/26563480565","builder_name":"checks-actions"} -->

workingjubilee commented:
gah.

@bors r=saethlin rollup=never

bors commented:
:pushpin: Commit 86a7efc7e1764f78c463a26701d027534376ad35 has been approved by `saethlin`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=saethlin 86a7efc7e1764f78c463a26701d027534376ad35 -->
<!-- homu: {"type":"Approved","sha":"86a7efc7e1764f78c463a26701d027534376ad35","approver":"saethlin","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

workingjubilee commented:
@bors r-

workingjubilee commented:
many targets to fix, and none of them are going to be part of this distinctly different PR...

@bors r=saethlin rollup=never

bors commented:
:pushpin: Commit ed9bb9f74e8c6eab5ed39f87a8d5859387416502 has been approved by `saethlin`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=saethlin ed9bb9f74e8c6eab5ed39f87a8d5859387416502 -->
<!-- homu: {"type":"Approved","sha":"ed9bb9f74e8c6eab5ed39f87a8d5859387416502","approver":"saethlin","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9631665171/job/26563763714) [(plain)](https://github.com/rust-lang/rust/commit/ed9bb9f74e8c6eab5ed39f87a8d5859387416502/checks/26563763714/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
 - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 46)
 - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 49)
 - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 61)
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
test result: ok. 292 passed; 0 failed; 5 ignored; 0 measured; 0 filtered out; finished in 4.70s

##[endgroup]
##[group]Testing stage2 compiletest suite=codegen mode=codegen (x86_64-unknown-linux-gnu)
error: detected unknown compiletest test directive `ignore-x86_64-apple-darwin` in /checkout/tests/codegen/force-frame-pointers.rs:9
errors encountered during EarlyProps parsing: /checkout/tests/codegen/force-frame-pointers.rs
errors encountered during EarlyProps parsing
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
Build completed unsuccessfully in 0:15:05
  local time: Sun Jun 23 07:33:09 UTC 2024
```

</details>

workingjubilee commented:
really? I thought I...
@bors r-

workingjubilee commented:
there... should have rebased so that the build worked again locally to start...

workingjubilee commented:
@bors r=saethlin

bors commented:
:pushpin: Commit c72a6ad8f2653d8da38a28520b6c5a2d64aa7aa5 has been approved by `saethlin`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=saethlin c72a6ad8f2653d8da38a28520b6c5a2d64aa7aa5 -->
<!-- homu: {"type":"Approved","sha":"c72a6ad8f2653d8da38a28520b6c5a2d64aa7aa5","approver":"saethlin","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit c72a6ad8f2653d8da38a28520b6c5a2d64aa7aa5 with merge 45af7d98aa0ffb1fbf10ca2277d16cbcec187b9d...
<!-- homu: {"type":"BuildStarted","head_sha":"c72a6ad8f2653d8da38a28520b6c5a2d64aa7aa5","merge_sha":"45af7d98aa0ffb1fbf10ca2277d16cbcec187b9d"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/9631821029/job/26565091222)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/9631821029/job/26565091222","builder_name":"checks-actions"} -->

workingjubilee committed:
compiler: parse the empty frame-pointer

workingjubilee commented:
fuchsia...

@bors r=saethlin

bors commented:
:pushpin: Commit 761ba5bb073578c6a510482be8338075f0b71493 has been approved by `saethlin`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=saethlin 761ba5bb073578c6a510482be8338075f0b71493 -->
<!-- homu: {"type":"Approved","sha":"761ba5bb073578c6a510482be8338075f0b71493","approver":"saethlin","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 761ba5bb073578c6a510482be8338075f0b71493 with merge c3d7fb398569407350abe044e786bc7890c90397...
<!-- homu: {"type":"BuildStarted","head_sha":"761ba5bb073578c6a510482be8338075f0b71493","merge_sha":"c3d7fb398569407350abe044e786bc7890c90397"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/9632989054/job/26568672780)
Approved by: saethlin
Pushing c3d7fb398569407350abe044e786bc7890c90397 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"saethlin","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/9632989054/job/26568672780"},"merge_sha":"c3d7fb398569407350abe044e786bc7890c90397"} -->

rust-timer commented:
Finished benchmarking commit (c3d7fb398569407350abe044e786bc7890c90397): [comparison URL](https://perf.rust-lang.org/compare.html?start=05468cf1241ae661c84e43d6a41d7a0ba317eebf&end=c3d7fb398569407350abe044e786bc7890c90397&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=05468cf1241ae661c84e43d6a41d7a0ba317eebf&end=c3d7fb398569407350abe044e786bc7890c90397&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=05468cf1241ae661c84e43d6a41d7a0ba317eebf&end=c3d7fb398569407350abe044e786bc7890c90397&stat=max-rss)
This benchmark run did not return any relevant results for this metric.

### [Cycles](https://perf.rust-lang.org/compare.html?start=05468cf1241ae661c84e43d6a41d7a0ba317eebf&end=c3d7fb398569407350abe044e786bc7890c90397&stat=cycles:u)
<details>
<summary>Results (secondary 2.3%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 2.3% | [2.1%, 2.6%] | 7     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=05468cf1241ae661c84e43d6a41d7a0ba317eebf&end=c3d7fb398569407350abe044e786bc7890c90397&stat=size:linked_artifact)
<details>
<summary>Results (secondary 0.0%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 0.0% | [0.0%, 0.0%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |
</details>

**Bootstrap**: 691.495s -> 693.516s (0.29%)
**Artifact size**: 326.88 MiB -> 326.77 MiB (-0.04%)
